File: rust/src/s0049.rs
use std::collections::HashMap;
pub struct Solution {}

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut map: HashMap<String, usize> = HashMap::new();
        let mut result = Vec::new();
        for s in strs {
            // create hash key
            // find if hash key in map
            // -> no: insert hash key in map, value is result vec len - 1
            // -> yes: get value from map, insert s to vec[value]
            let key = Self::get_key(&s);
            if !map.contains_key(&key) {
                result.push(vec![s]);
                map.insert(key, result.len() - 1);
            } else {
                result[map[&key]].push(s);
            }
        }
        result
    }

    fn get_key(s: &String) -> String {
        let mut chars: Vec<char> = s.chars().collect();
        chars.sort();
        chars.iter().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        Solution {};
    }
}

File: cpp/s0049.cc
class Solution {
public:
  vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> map;
    for(string &str: strs) {
      array<int, 26> count{};
      for (char &c: str) {
        count[c - 'a'] += 1;
      }
      map[get_key(count)].emplace_back(str);
    }
    vector<vector<string>> result;
    for(auto &[key, values]: map) {
      result.emplace_back(values);
    }
    return result;
  }

  string get_key(array<int, 26> &count) {
    string key = "";
    for(int &c: count) {
      key += to_string(c) + ",";
    }
    return key;
  }
};
